### ISSUE TITLE
Style/GuardClause suggests changes that are impossible due variable scope

### ISSUE THREAD

**Issue by drbrain** [rubocop-hq/rubocop#4581]
## Expected behavior

The Style/GuardClause cop does not suggest you "Use a guard clause" that results in a NoMethodError.

Instead the Style/GuardClause cop should suggest a change that doesn't result in a NoMethodError, something like "separate your conditional into multiple lines and use a guard clause"

## Actual behavior

The Style/GuardClause cop suggests "[using] a guard clause" that results in a NoMethodError

## Steps to reproduce the problem

Here is a working program that triggers Style/GuardClause:

```
$ cat t.rb
def m value
  if match = /(?<hi>hi)/.match(value)
    match[:hi]
  end
end

p m 'hi'
p m 'bye'
$ ruby -v t.rb
ruby 2.4.1p111 (2017-03-22 revision 58053) [x86_64-darwin16]
"hi"
nil
```

Rubocop output suggests I "Use a guard clause":

```
$ rubocop --only Style/GuardClause t.rb
Inspecting 1 file
C

Offenses:

t.rb:2:3: C: Use a guard clause instead of wrapping the code inside a conditional expression.
  if match = /(?<hi>hi)/.match(value)
  ^^

1 file inspected, 1 offense detected
```

Let's remove the guard clause:

```
$ cat t.rb
def m value
  match[:hi] if match = /(?<hi>hi)/.match(value)
end

p m 'hi'
p m 'bye'
```

Rubocop is OK:

```
$ rubocop --only Style/GuardClause t.rb
Inspecting 1 file
.

1 file inspected, no offenses detected
```

Ruby is not:

```
$ ruby t.rb
t.rb:2:in `m': undefined local variable or method `match' for main:Object (NameError)
Did you mean?  catch
	from t.rb:5:in `<main>'
```

Ruby parsers strictly left-to-right so the first time `match` is encountered there is no `match` variable because the assignment is further right in the conditional.  `match` is considered a method at that point, but there is no `match` method so we get a NoMethodError upon execution.

There is no way to "Use a guard clause" here, you must make a larger change:

```
$ cat t.rb
def m value
  match = /(?<hi>hi)/.match(value)

  match[:hi] if match
end

p m 'hi'
p m 'bye'
```

Rubocop should suggest something like "separate your conditional into multiple lines and use a guard clause"

## RuboCop version

Include the output of `rubocop -V`. Here's an example:

```
$ rubocop -V
0.49.1 (using Parser 2.4.0.0, running on ruby 2.4.1 x86_64-darwin16)
```

**Comment by Drenmi** [comment]
I think RuboCop suggest doing this:

```ruby
def m value
  return unless match = /(?<hi>hi)/.match(value)

  match[:hi]
end
```

Maybe it's possible to make the message more specific somehow?

**Comment by drbrain** [comment]
One of the most difficult things new rubyists seem to encounter is the behavior difference of the modifier-`if` with respect to variable assignment (per above), so rubocop needs to be extra clear here because new rubyists will not know why they get `NoMethodError` and will be confused when they get it.

If `rubocop -a` can't fix it then the message should be clear enough that a new rubyist knows what to do.

**Comment by bbatsov** [comment]
No arguments here! PRs to improve the message are welcome!
